README.md
azure-queue-node
================

Simplified Azure Queue Storage client library for NodeJS.

What is supported:

* creating, deleting and listing queues
* putting, updating, getting and deleting messages
* retrieve approximate number of messages in queue

What may added in the future:

* generating SAS (Shared Access Signature) and using it for authentication
* get and set service properties
* get and set per queue metadata

Usage
==============

## Setting the default client connection info and other settings

Default client uses environment variable to set up the access key and storage URL if possible. It looks for the `CLOUD_STORAGE_ACCOUNT` setting with three elements (it is the default format used by Azure Storage):

```
QueueEndpoint=http://accountName.queue.core.windows.net/;AccountName=accountName;AccountKey=theKey
```

No error is returned if this doesn't exists, is incomplete or malformed.

*Current version does not support quotes and AccountKey must be the last one to be parsed correctly. This may be fixed in the future.*

If the environment variable is not set, the default connection info have to be set using below command to be usable:

```javascript
var azureQueue = require('azure-queue-node');
azureQueue.setDefaultClient({
  accountUrl: 'http://accountName.queue.core.windows.net/',
  accountName: 'accountName',
  accountKey: 'theKey'
});
```

The same method allows to set other default client settings (see *Client settings*).

## Using default client

```javascript
var azureQueue = require('azure-queue-node');
var defaultClient = azureTable.getDefaultClient();

// use the client to create the queue
defaultClient.createQueue('queueName', true, cb);
// or add message
defaultClient.putMessage('queue', message, options, cb);
```

## Creating customized client

It is possible to create additional clients that are based on other client (or on default settings), but customized and independent. This allows to for example use several queue storage accounts but still have one default for convenience.

```javascript
var azureQueue = require('azure-queue-node');
var queueClient = azureQueue.createClient({
  // predefined settings
}, [baseClient]);
```

Base client is the client on which the new one will be based. If not provided, it is based on the default one.

## Longer usage example

```javascript
var client = azureQueue.getDefaultClient();
client.putMessage('taskqueue', {
  value1: 'ABCDEFG'
}, function(err, data) {
  // err is null
  // data is undefined
});
client.getMessages('taskqueue', {maxMessages: 10}, function(err, data) {
  // err is null
  // data contains array with to 10 queue message objects
});
client.deleteMessage('taskqueue', 'messageId', 'popReceipt', function(err, data) {
  // err is null
  // data is undefined
});
```

Client settings
===============

Account related:

* `accountUrl` (string) - URL of the service's endpoint (no default value)
* `accountName` (string) - name of the used account (no default value)
* `accountKey` (string) - base64 encoded account key (no default value), may be null if `sas` is provided

Underlying HTTP request related (passed without changes to request module):

* `timeout` (int) - request timeout in miliseconds (default: 30000)
* `proxy` (string) - proxy URL
* `forever` (bool) - use `true` to turn advanced socket reuse
* `agent` (Agent) - already created agent object (do not set `proxy`, `forever` or `pool` is set (may not work otherwise)
* `agentOptions` (object) - used to set maxSockets option for forever or standard agent
* `pool` (false|object) - use `false` to turn off socket reuse

Azure Queue Storage related:

* `json` (bool) - if set to `true` (default), message is JSON encoded/decoded; otherwise it is base64 encoded/decoded
* `base64` (bool) - if set to `true` (`false` by default) forces JSON message to be base64 encoded (for compatibility if using older clients)
* `retry` (false/object/function) - set to `false` to turn off any retry policy; provide a function for custom retry logic or use object to change parameter of build in retry logic

Retry options:

* `retries` (int) - a number of retries (default: 3)
* `firstDelay` (int) - delay of the first retry request in ms (default: 2000ms)
* `nextDelayMult` (float) - delay multiplier using previous delay as a base (default: 2); use 1 for linear delay
* `variability` (float) - random delay multiplier added or subtracted from current delay (default 0.2)
* `transientErrors` (array of ints or strings) - describes situations where retry should be used; if it is int, status code is checked for equality; for string the code element of error or response is checked (default: `[500, 501, 502, 503, 'ETIMEDOUT', 'ECONNRESET', 'EADDRINUSE', 'ESOCKETTIMEDOUT']`

Custom retry function should support below parameters (in order of appearance):

* `requestOptions` (object) - if needed it can be edited in place to change request options and headers
* `nextReq` (function) - function to be called to make a request.

The `nextReq` function must be called passing additional function `retryFn` with below parameters:

* `err` (object) - null or error object from response (see normal response callback)
* `resp` (object/array) - response object (see normal response callback)
* `nextResp` (function) - function to be called when there is no need for a retry passing (`err` and `resp` to it)

As by default request headers are not regenerated on retries, the retry time of of the last one cannot be very long, because authentication will fail.

Example retry function which retries every time `code` in error is `ETIMEDOUT`:

```javascript
function _retryLogic(requestOptions, nextReq) {
  function retryFn(err, resp, nextResp) {
    if (err && err.code === 'ETIMEDOUT') {
      nextReq(retryFn);
    } else {
      nextResp(err, resp);
    }
  }
  nextReq(retryFn);
}
```

In case of reties, the default retry function returns `retriesMade` key in error informing about the number of retries made (helps in debugging).

API
===

If not explained differently, `cb` in API is a functions in format `function cb(err, data)`, but `data` may be undefined if there is no explicit response. In case of queue list there may be additional third argument passed if there is a marker token.

If not stated otherwise, `options` parameter may contain `timeout` key with timeout in seconds specific for this operation. If not provided, Azure Storage uses its default (30 seconds).

## Module level

### getDefaultClient()

Returns default `Client` object. If `setDefaultClient()` was not used earlier the client only have default module settings and environment variable applied.

### setDefaultClient(settings)

Sets up and returns default `Client` object with provided `settings`. It is using default settings as a base.

### createClient(settings, [base])

Returns new `Client` object using new settings and `base` client settings as a fallback. If `base` is not provided, uses default client settings.


## Client object level

### create(settings)

Returns new `Client` object using only provided settings object. Shouldn't be used directly unless you want to provide all options. Use `createClient` from main module if possible.

### getSettings()

Returns sealed settings object used by this client.

### createQueue(queue, [options], cb)

Creates new queue. The `queue` is queue name. The `options` is optional and supports only `timeout` key. The `cb` is a standard callback function, but `data` is `true` if queue was created and `false` if already existed.

### deleteQueue(queue, [options], cb)

Removes existing queue. The `queue` is queue name. The `cb` is a standard callback function.

### listQueues([options], cb)

Returns array with queue names (as strings). The `options` is optional, but if exists and `marker` key is provided, the retrieval will start from last marker token. The `prefix` in `options` allow to filter the results to return only queues with names that begin with the specified prefix. The `maxResults` in `options` allow to retrieve lower about of results. The `cb` is a standard callback function, but if marker is returned, the third argument will be passed with value for `marker` key.

### countMessages(queue, [options], cb)

Returns approximate queue messages count. The `queue` is queue name. The `cb` is a standard callback function. The `data` part will contain integer with approximate count of messages in queue.

### putMessage(queue, message, [options], cb)

Add new `message` to `queue`. The `message` may be a base64 string (if JSON is not used) or object (it will be JSON encoded). The `options` have two time keys supported: `visibilityTimeout` and `messageTTL`; both are times in seconds (see Azure documentation). In addition `options` allow to override `json` setting per request.

### clearMessages(queue, [options], cb)

Removes messages from queue. The `queue` is queue name. The `cb` is a standard callback function.

### getMessages(queue, [options], cb)

Retrieves messages from queue. The `messageText` may be decoded to string (from bas64 if JSON is not used) or object (if was JSON encoded). The `options` have two main keys supported: `visibilityTimeout` (in seconds;see Azure documentation) and `maxMessages` (number of messages to retrieve; max is 32, default is 1). In addition `options` allow to override `json` setting per request.

The `data` in callback is always an array (even if there is no messages in queue). Each message may have below fields:

```
{
"messageId": "f00c98a4-1547-4d65-a5ea-ec233a85d7af",
"popReceipt": "sK52prNk0QgBAAAA",
"messageText": {}, // object or string depending on settings
"dequeueCount": integer, // number of gets without delete
"expirationTime": Date object
"insertionTime": Date object
"timeNextVisible": Date object
}
```

### peekMessages(queue, [options], cb)

Retrieves messages from queue. The `messageText` may be decoded to string (from bas64 if JSON is not used) or object (if was JSON encoded). The `options` supports `maxMessages` (number of messages to retrieve; max is 32, default is 1). In addition `options` allow to override `json` setting per request.

The `data` in callback is always an array (even if there is no messages in queue). Each message may have below fields:

```
{
"messageId": "f00c98a4-1547-4d65-a5ea-ec233a85d7af",
"messageText": {}, // object or string depending on settings
"dequeueCount": integer, // number of gets without delete
"expirationTime": Date object
"insertionTime": Date object
}
```

### updateMessage(queue, messageId, popReceipt, visibilityTimeout, message, [options], cb)

Updates `message` in `queue`. The `messageId` and `popReceipt` are values read using `getMessages()` or `popMessages()`. The `message` may be a base64 string (if JSON is not used) or object (it will be JSON encoded). The new `visibilityTimeout` is expressed in seconds (see Azure documentation). In addition `options` allow to override `json` setting per request.

The `data` object returned in callback contains new values of `timeNextVisible` (Date object) and `popReceipt`.

### deleteMessage(queue, messageId, popReceipt, [options], cb)

Removes `messageId` with `popReceipt` from `queue`. The `options` object is optional. The `data` in callback will always be undefined.


Running tests
=============

Run the tests using mocha from main project folder. But before that set the environment variable as some tests are relying on default:

```
set CLOUD_STORAGE_ACCOUNT=QueueEndpoint=http://127.0.0.1:10001/devstoreaccount1/;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==
mocha
```

CHANGELOG.txt
1.1.0
Added base64 encoding to JSON objects (as some Queue clients only support base64). By default is set to false.
Fixed base64 decoding in some cases.

1.0.2
Fixed clearMessages() function -- options param was in DOCS, but weren't supported in code.

1.0.1
Decode a base64 to object if JSON is expected (to allow using VS2013 queue browser in testing or Azure Scheduler queue)

1.0.0
Initial release

test/test-queues.js
/* jshint expr: true, quotmark: false */
/* globals it, describe, before, afterEach */
'use strict';

var expect = require('chai').expect;
var nock = require('nock');
var azureQueue = require('../index');

//nock.recorder.rec();

describe('default client', function() {
  var client;

  before(function(){
    client = azureQueue.getDefaultClient();
  });

  afterEach(function(){
    nock.cleanAll();
  });

  it('should create testqueue queue using required headers', function(done) {
    var azure = nock('http://127.0.0.1:10001')
      .matchHeader('x-ms-version', '2014-02-14')
      .matchHeader('authorization', /SharedKey/i)
      .matchHeader('x-ms-date', / \d\d:\d\d:\d\d /)
      .put('/devstoreaccount1/testqueue')
      .reply(201, "", { 'transfer-encoding': 'chunked',
        server: 'Windows-Azure-Queue/1.0 Microsoft-HTTPAPI/2.0',
        'x-ms-request-id': '34ecf932-5706-43ca-83a9-eecdce111c3f',
        'x-ms-version': '2014-02-14',
        date: 'Wed, 02 Jul 2014 09:08:07 GMT' });

    client.createQueue('testqueue', function(err, data) {
      expect(err).to.be.null;
      expect(data).to.be.true;
      expect(azure.isDone()).to.be.true;

      done();
    });
  });

  it('should not return error when testqueue is already created', function(done) {
    var azure = nock('http://127.0.0.1:10001')
      .put('/devstoreaccount1/testqueue')
      .reply(204, "", { 'content-length': '0',
        server: 'Windows-Azure-Queue/1.0 Microsoft-HTTPAPI/2.0',
        'x-ms-request-id': 'ae05573d-dc7a-41de-8695-171fe8b3618e',
        'x-ms-version': '2014-02-14',
        date: 'Wed, 02 Jul 2014 09:35:47 GMT' });

    client.createQueue('testqueue', function(err, data) {
      expect(err).to.be.null;
      expect(data).to.be.false;
      expect(azure.isDone()).to.be.true;

      done();
    });
  });

  it('should return a list of queues without any additions', function(done) {
    var azure = nock('http://127.0.0.1:10001')
      .get('/devstoreaccount1/?comp=list')
      .reply(200, "﻿<?xml version=\"1.0\" encoding=\"utf-8\"?><EnumerationResults ServiceEndpoint=\"http://127.0.0.1:10001/devstoreaccount1/\"><Queues><Queue><Name>aqueue1</Name></Queue><Queue><Name>aqueue2</Name></Queue><Queue><Name>testqueue</Name></Queue></Queues><NextMarker /></EnumerationResults>", { 'cache-control': 'no-cache',
        'transfer-encoding': 'chunked',
        'content-type': 'application/xml',
        server: 'Windows-Azure-Queue/1.0 Microsoft-HTTPAPI/2.0',
        'x-ms-request-id': '26fe5ecf-09f6-47ef-ae36-38f9d00c5b51',
        'x-ms-version': '2014-02-14',
        date: 'Wed, 02 Jul 2014 10:25:02 GMT' });

    client.listQueues(function(err, data, marker) {
      expect(err).to.be.null;
      expect(data).to.be.deep.equal(['aqueue1', 'aqueue2', 'testqueue']);
      expect(marker).to.be.undefined;
      expect(azure.isDone()).to.be.true;

      done();
    });
  });

  it('should return a list of queues with marker using maxResults and prefix', function(done) {
    var azure = nock('http://127.0.0.1:10001')
      .get('/devstoreaccount1/?comp=list&prefix=aq&maxresults=1')
      .reply(200, "﻿<?xml version=\"1.0\" encoding=\"utf-8\"?><EnumerationResults ServiceEndpoint=\"http://127.0.0.1:10001/devstoreaccount1/\"><Prefix>aq</Prefix><MaxResults>1</MaxResults><Queues><Queue><Name>aqueue1</Name></Queue></Queues><NextMarker>aqueue2</NextMarker></EnumerationResults>", { 'cache-control': 'no-cache',
        'transfer-encoding': 'chunked',
        'content-type': 'application/xml',
        server: 'Windows-Azure-Queue/1.0 Microsoft-HTTPAPI/2.0',
        'x-ms-request-id': '984f3644-4010-4b2b-8f50-b5c063deb029',
        'x-ms-version': '2014-02-14',
        date: 'Wed, 02 Jul 2014 10:18:10 GMT' });

    client.listQueues({
      prefix: 'aq',
      maxResults: 1
    }, function(err, data, marker) {
      expect(err).to.be.null;
      expect(data).to.be.an.array;
      expect(data).to.have.length(1);
      expect(data[0]).to.be.equal('aqueue1');
      expect(marker).to.be.equal('aqueue2');
      expect(azure.isDone()).to.be.true;

      done();
    });
  });

  it('should return a list using marker and timeout', function(done) {
    var azure = nock('http://127.0.0.1:10001')
      .get('/devstoreaccount1/?comp=list&marker=aqueue2&timeout=10')
      .reply(200, "﻿<?xml version=\"1.0\" encoding=\"utf-8\"?><EnumerationResults ServiceEndpoint=\"http://127.0.0.1:10001/devstoreaccount1/\"><Marker>aqueue2</Marker><Queues><Queue><Name>aqueue2</Name></Queue><Queue><Name>testqueue</Name></Queue></Queues><NextMarker /></EnumerationResults>", { 'cache-control': 'no-cache',
        'transfer-encoding': 'chunked',
        'content-type': 'application/xml',
        server: 'Windows-Azure-Queue/1.0 Microsoft-HTTPAPI/2.0',
        'x-ms-request-id': '73f28ca9-ad77-4f1f-be3b-a449e45d13ae',
        'x-ms-version': '2014-02-14',
        date: 'Wed, 02 Jul 2014 11:12:10 GMT' });

    client.listQueues({
      marker: 'aqueue2',
      timeout: 10
    }, function(err, data, marker) {
      expect(err).to.be.null;
      expect(data).to.be.deep.equal(['aqueue2', 'testqueue']);
      expect(marker).to.be.undefined;
      expect(azure.isDone()).to.be.true;

      done();
    });
  });

  it('should return empty list of there is no queues with prefix', function(done) {
    var azure = nock('http://127.0.0.1:10001')
      .get('/devstoreaccount1/?comp=list&prefix=zebra')
      .reply(200, "﻿<?xml version=\"1.0\" encoding=\"utf-8\"?><EnumerationResults ServiceEndpoint=\"http://127.0.0.1:10001/devstoreaccount1/\"><Prefix>zebra</Prefix><Queues /><NextMarker /></EnumerationResults>", { 'cache-control': 'no-cache',
        'transfer-encoding': 'chunked',
        'content-type': 'application/xml',
        server: 'Windows-Azure-Queue/1.0 Microsoft-HTTPAPI/2.0',
        'x-ms-request-id': '8d5fcc72-2300-4278-9a17-7131003b0bd4',
        'x-ms-version': '2014-02-14',
        date: 'Wed, 02 Jul 2014 11:15:37 GMT' });

    client.listQueues({
      prefix: 'zebra'
    }, function(err, data, marker) {
      expect(err).to.be.null;
      expect(data).to.be.deep.equal([]);
      expect(marker).to.be.undefined;
      expect(azure.isDone()).to.be.true;

      done();
    });
  });

  it('should remove queue without errors', function(done) {
    var azure = nock('http://127.0.0.1:10001')
      .delete('/devstoreaccount1/aqueue1')
      .reply(204, "", { 'content-length': '0',
        server: 'Windows-Azure-Queue/1.0 Microsoft-HTTPAPI/2.0',
        'x-ms-request-id': 'cbb400e2-2cb4-4e84-a21f-e0761732514b',
        'x-ms-version': '2014-02-14',
        date: 'Wed, 02 Jul 2014 11:44:05 GMT' });

    client.deleteQueue('aqueue1', function(err, data) {
      expect(err).to.be.null;
      expect(data).to.be.undefined;

      expect(azure.isDone()).to.be.true;

      done();
    });
  });

  it('should return error if queue is removed', function(done) {
    var azure = nock('http://127.0.0.1:10001')
      .delete('/devstoreaccount1/aqueue1')
      .reply(404, "﻿<?xml version=\"1.0\" encoding=\"utf-8\"?><Error><Code>QueueNotFound</Code><Message>The specified queue does not exist.\nRequestId:052e5b46-2117-43d3-a33c-ce296adb7a72\nTime:2014-07-02T11:53:02.7276220Z</Message></Error>", { 'content-length': '217',
        'content-type': 'application/xml',
        server: 'Windows-Azure-Queue/1.0 Microsoft-HTTPAPI/2.0',
        'x-ms-request-id': '052e5b46-2117-43d3-a33c-ce296adb7a72',
        'x-ms-version': '2014-02-14',
        date: 'Wed, 02 Jul 2014 11:53:02 GMT' });

    client.deleteQueue('aqueue1', function(err, data) {
      expect(err).to.not.be.null;
      expect(data).to.be.undefined;
      expect(azure.isDone()).to.be.true;

      expect(err).to.have.property('statusCode', 404);
      expect(err).to.have.property('code', 'QueueNotFound');

      done();
    });
  });

  it('should return approximate queue length', function(done) {
    var azure = nock('http://127.0.0.1:10001')
      .head('/devstoreaccount1/testqueue?comp=metadata')
      .reply(200, "", { 'content-length': '0',
        server: 'Windows-Azure-Queue/1.0 Microsoft-HTTPAPI/2.0',
        'x-ms-request-id': 'cbb400e2-2cb4-4e84-a21f-e0761732514b',
        'x-ms-version': '2014-02-14',
        'x-ms-approximate-messages-count': '5',
        date: 'Wed, 02 Jul 2014 11:44:05 GMT' });

    client.countMessages('testqueue', function(err, data) {
      expect(err).to.be.null;
      expect(data).to.be.equal(5);
      expect(azure.isDone()).to.be.true;

      done();
    });
  });

});

lib/client.js
'use strict';

var request = require('request');
var url = require('url');
var querystring = require('querystring');
var xml2js = require('xml2js');

var versionInfo = require('../package.json').version;
var utils = require('./utils');

var XML_PARSER_OPTIONS = {
  ignoreAttrs: true,
  trim: true,
  tagNameProcessors: [xml2js.processors.firstCharLowerCase],
  explicitArray: false
};

var emptyObject = {};

var Client = {
  // settings object, cannot be edited
  _settings: null,
  // request object with defaults for this client
  _request: null,
  // decoded azure key
  _azureKey: null,
  // retry function
  _retryLogic: null,

  _prepareRequestDefaults: function(settings) {
    var defaults = {
      encoding: 'utf8',
      timeout: settings.timeout
    };
    if (settings.agent) {
      defaults.agent = settings.agent;
    }
    if (settings.proxy) {
      defaults.proxy = settings.proxy;
    }
    if (settings.forever === true) {
      defaults.forever = settings.forever;
    }
    if (settings.agentOptions) {
      defaults.agentOptions = settings.agentOptions;
    }
    if (settings.pool != null) {
      defaults.pool = settings.pool;
    }
    return defaults;
  },

  _getRequestSpecificOptions: function _getRequestSpecificOptions(method, path, qs, body) {
    var now = new Date().toUTCString();

    var requestOptions = {
      method: method,
      uri: url.parse(this._settings.accountUrl + path),
      qs: qs,
      headers: {
        date: now,
        'user-agent': 'azure-queue-node/'+versionInfo,
        'content-type': 'application/xml;charset="utf-8"',
        'content-length': body ? Buffer.byteLength(body, 'utf8') : 0,
        'x-ms-date': now,
        'x-ms-version': '2014-02-14'
      }
    };

    return requestOptions;
  },

  _addSharedKeyAuthHeader: function _addSharedKeyAuthHeader(requestOptions) {
    var stringToSign = requestOptions.method +'\n';
    stringToSign += (requestOptions.headers['content-encoding'] ? requestOptions.headers['content-encoding'] : '') + '\n';
    stringToSign += (requestOptions.headers['content-language'] ? requestOptions.headers['content-language'] : '') + '\n';
    stringToSign += (requestOptions.headers['content-length'] != null ? requestOptions.headers['content-length'] : '') + '\n';
    stringToSign += (requestOptions.headers['content-md5'] ? requestOptions.headers['content-md5'] : '') + '\n';
    stringToSign += (requestOptions.headers['content-type'] ? requestOptions.headers['content-type'] : '') + '\n';
    //stringToSign += (requestOptions.headers['date'] ? requestOptions.headers['date'] : '') + '\n';
    stringToSign += '\n'; // do not count date header (Microsoft's strange logic!)
    // If-Modified-Since, If-Match, If-None-Match, If-Unmodified-Since, Range are not used, so can be skipped
    stringToSign += '\n\n\n\n\n';
    // CanonicalizedHeaders, only below ones, as currently lib doesn't support queue metadata
    stringToSign += (requestOptions.headers['x-ms-date'] ? 'x-ms-date:'+requestOptions.headers['x-ms-date']+'\n' : '');
    stringToSign += (requestOptions.headers['x-ms-version'] ? 'x-ms-version:'+requestOptions.headers['x-ms-version']+'\n' : '');
    // CanonicalizedResource
    stringToSign += '/'+this._settings.accountName;
    stringToSign += requestOptions.uri.path;
    // as possible query string elements are very limited, ifs inlined
    if (requestOptions.qs) {
      if ('comp' in requestOptions.qs) {
        stringToSign += '\ncomp:' + requestOptions.qs.comp;
      }
      if ('marker' in requestOptions.qs) {
        stringToSign += '\nmarker:' + requestOptions.qs.marker;
      }
      if ('maxresults' in requestOptions.qs) {
        stringToSign += '\nmaxresults:' + requestOptions.qs.maxresults;
      }
      if ('messagettl' in requestOptions.qs) {
        stringToSign += '\nmessagettl:' + requestOptions.qs.messagettl;
      }
      if ('numofmessages' in requestOptions.qs) {
        stringToSign += '\nnumofmessages:' + requestOptions.qs.numofmessages;
      }
      if ('peekonly' in requestOptions.qs) {
        stringToSign += '\npeekonly:' + requestOptions.qs.peekonly;
      }
      if ('prefix' in requestOptions.qs) {
        stringToSign += '\nprefix:' + requestOptions.qs.prefix;
      }
      if ('popreceipt' in requestOptions.qs) {
        stringToSign += '\npopreceipt:' + requestOptions.qs.popreceipt;
      }
      if ('timeout' in requestOptions.qs) {
        stringToSign += '\ntimeout:' + requestOptions.qs.timeout;
      }
      if ('visibilitytimeout' in requestOptions.qs) {
        stringToSign += '\nvisibilitytimeout:' + requestOptions.qs.visibilitytimeout;
      }
    }
    requestOptions.headers.authorization = 'SharedKey ' + this._settings.accountName + ':' + utils.hmacSha256(this._azureKey, stringToSign);
    return requestOptions;
  },

  _normalizeCallback: function _normalizeCallback(cb, error, response, body) {
    if (error) {
      return cb(error);
    }
    if (!response) {
      return cb({code: 'UnknownError'});
    }

    // check if response is really XML, if yes, parse it
    if (body && typeof body === 'string' && body.indexOf('<?xml' === 0)) {
      xml2js.parseString(body, XML_PARSER_OPTIONS, function(err, result) {
        if (err) {
          return cb({code: 'XMLParseError'});
        }
        body = result;
        next();
      })
    } else {
      next();
    }

    function next() {
      if (response.statusCode >= 400) {
        return cb({
          statusCode: response.statusCode,
          code: body && body.error ? body.error.code : 'UnknownBody',
          body: body && body.error ? body.error : body
        });
      }
      return cb(null, {
        statusCode: response.statusCode,
        headers: response.headers, // some important data may be in response headers
        body: body
      });
    }
  },

  _sendRequestWithRetry: function _sendRequestWithRetry(options, cb) {
    if (this._retryLogic == null) {
      this._request(options, this._normalizeCallback.bind(this, cb));
    } else {
      var self = this;
      this._retryLogic(options, function(filterCb) {
        self._request(options, self._normalizeCallback.bind(self, function(err, resp) {
          filterCb(err, resp, function(err, resp) {
            cb(err, resp);
          });
        }));
      });
    }
  },

  _makeRequest: function _makeRequest(method, path, qs, body, cb) {
    var options = this._getRequestSpecificOptions(method, path, qs, body);
    if (this._azureKey) {
      options = this._addSharedKeyAuthHeader(options);
    }

    if (typeof body === 'string' || typeof body === 'object') {
      options.body = body;
    }

    this._sendRequestWithRetry(options, cb);
  },

  create: function create(settings) {
    if (!settings.accountUrl || !settings.accountName || !settings.accountKey) {
      throw 'Provide accountUrl, accountName, and accountKey in settings or in env CLOUD_STORAGE_ACCOUNT';
    }

    var sealedSettings = Object.seal(settings);

    // create request object with most of the default settings
    var defaultRequest = request.defaults(this._prepareRequestDefaults(sealedSettings));

    var retryLogic;
    if (typeof sealedSettings.retry === 'function') {
      retryLogic = sealedSettings.retry;
    } else if (typeof sealedSettings.retry === 'object') {
      retryLogic = utils.generateRetryFunction(sealedSettings.retry);
    } else if (sealedSettings.retry === false) {
      retryLogic = null;
    } else {
      retryLogic = utils.generateRetryFunction();
    }

    return Object.create(this, {
      _settings: {value: sealedSettings},
      _request: {value: defaultRequest},
      _azureKey: {value: sealedSettings.accountKey ? utils.base64Decode(sealedSettings.accountKey) : null},
      _retryLogic: {value: retryLogic}
    });
  },

  getSettings: function getSettings() {
    return this._settings;
  },

  _createQueueCb: function _createQueueCb(cb, options, err, data) {
    if (!err && (data.statusCode === 201 || data.statusCode === 204)) {
      return cb(null, data.statusCode === 201);
    } else {
      return cb(err);
    }
  },
  createQueue: function createQueue(queueName, options, cb) {
    if (typeof options === 'function') {
      cb = options;
    }
    var qs = null;
    if (typeof options === 'object' && options.timeout) {
      qs = {
        timeout: options.timeout
      };
    }
    this._makeRequest('PUT', queueName, qs, null, this._createQueueCb.bind(this, cb, typeof options === 'object' ? options : null));
    return this;
  },

  _listQueuesCb: function _listQueuesCb(cb, err, data) {
    if (!err && data.statusCode === 200) {
      var er = data.body.enumerationResults;
      var results = new Array(er.queues.queue ? er.queues.queue.length || 1 : 0);
      if (results.length === 1) {
        results[0] = er.queues.queue.name;
      } else if (results.length > 1) {
        er.queues.queue.forEach(function(r, i) {
          this[i] = r.name;
        }, results);
      }
      var continuation = er.nextMarker ? er.nextMarker : undefined;
      return cb(null, results, continuation);
    } else {
      return cb(err);
    }
  },
  listQueues: function listQueues(options, cb){
    if (typeof options === 'function') {
      cb = options;
    }
    var qs = {comp: 'list'};
    if (typeof options === 'object') {
      if (typeof options.prefix === 'string') {
        qs.prefix = options.prefix;
      }
      if (typeof options.marker === 'string') {
        qs.marker = options.marker;
      }
      if (options.maxResults) {
        qs.maxresults = options.maxResults;
      }
      if (options.timeout) {
        qs.timeout = options.timeout;
      }
    }

    this._makeRequest('GET', '', qs, null, this._listQueuesCb.bind(this, cb));
    return this;
  },

  _204Cb: function _204Cb(cb, err, data) {
    if (!err && data.statusCode === 204) {
      return cb(null);
    } else {
      return cb(err);
    }
  },

  deleteQueue: function deleteQueue(queueName, options, cb) {
    if (typeof options === 'function') {
      cb = options;
    }
    var qs = null;
    if (typeof options === 'object' && options.timeout) {
      qs = {
        timeout: options.timeout
      };
    }
    this._makeRequest('DELETE', queueName, qs, null, this._204Cb.bind(this, cb));
    return this;
  },

  _countMessagesCb: function _countMessagesCb(cb, err, data) {
    if (!err && data.statusCode === 200) {
      return cb(null, parseInt(data.headers['x-ms-approximate-messages-count'], 10));
    } else {
      return cb(err);
    }
  },
  countMessages: function countMessages(queueName, options, cb) {
    if (typeof options === 'function') {
      cb = options;
    }
    var qs = {
      comp: 'metadata'
    };
    if (typeof options === 'object' && options.timeout) {
      qs = {
        timeout: options.timeout
      };
    }
    this._makeRequest('HEAD', queueName, qs, null, this._countMessagesCb.bind(this, cb));
    return this;
  },

  _putMessageCb: function _putMessageCb(cb, err, data) {
    if (!err && data.statusCode === 201) {
      return cb(null);
    } else {
      return cb(err);
    }
  },
  putMessage: function putMessage(queue, message, options, cb) {
    if (typeof options === 'function') {
      cb = options;
    }

    var qs = null;
    if (typeof options === 'object') {
      qs = {};
      if (options.visibilityTimeout) {
        qs.visibilitytimeout = options.visibilityTimeout;
      }
      if (options.messageTTL) {
        qs.messagettl = options.messageTTL;
      }
      if (options.timeout) {
        qs.timeout = options.timeout;
      }
      if (Object.keys(qs).length === 0) {
        qs = null;
      }
    }

    this._makeRequest('POST', queue+'/messages', qs, utils.xmlizeMessage.call(this, options || emptyObject, message), this._putMessageCb.bind(this, cb));
    return this;
  },

  clearMessages: function clearMessages(queueName, options, cb) {
    if (typeof options === 'function') {
      cb = options;
    }
    var qs = null;
    if (typeof options === 'object' && options.timeout) {
      qs = {
        timeout: options.timeout
      };
    }
    this._makeRequest('DELETE', queueName+'/messages', qs, null, this._204Cb.bind(this, cb));
    return this;
  },

  _getMessagesCb: function _getMessagesCb(cb, options, err, data) {
    if (!err && data.statusCode === 200) {
      var er = data.body.queueMessagesList;
      var results = new Array(er.queueMessage ? er.queueMessage.length || 1 : 0);
      if (results.length === 1) {
        er.queueMessage = utils.decodeMessage.call(this, options || emptyObject, er.queueMessage);
        results[0] = er.queueMessage;
      } else if (results.length > 1) {
        var self = this;
        er.queueMessage.forEach(function(r, i) {
          this[i] = utils.decodeMessage.call(self, options || emptyObject, r);
        }, results);
      }
      return cb(null, results);
    } else {
      return cb(err);
    }
  },
  getMessages: function getMessages(queue, options, cb) {
    if (typeof options === 'function') {
      cb = options;
    }

    var qs = null;
    if (typeof options === 'object') {
      qs = {};
      if (options.visibilityTimeout) {
        qs.visibilitytimeout = options.visibilityTimeout;
      }
      if (options.maxMessages) {
        qs.numofmessages = options.maxMessages;
      }
      if (options.timeout) {
        qs.timeout = options.timeout;
      }
      if (Object.keys(qs).length === 0) {
        qs = null;
      }
    }

    this._makeRequest('GET', queue+'/messages', qs, null, this._getMessagesCb.bind(this, cb, options));
    return this;
  },

  peekMessages: function peekMessages(queue, options, cb) {
    if (typeof options === 'function') {
      cb = options;
    }

    var qs = {
      peekonly: 'true'
    };
    if (options.maxMessages) {
      qs.numofmessages = options.maxMessages;
    }
    if (options.timeout) {
      qs.timeout = options.timeout;
    }

    // we can use get _getMessagesCb here
    this._makeRequest('GET', queue+'/messages', qs, null, this._getMessagesCb.bind(this, cb, options));
    return this;
  },

  _updateMessageCb: function _updateMessageCb(cb, err, data) {
    if (!err && data.statusCode === 204) {
      return cb(null, {
        popReceipt: data.headers['x-ms-popreceipt'],
        timeNextVisible: new Date(data.headers['x-ms-time-next-visible'])
      });
    } else {
      return cb(err);
    }
  },
  updateMessage: function updateMessage(queue, messageId, popReceipt, visibilityTimeout, message, options, cb) {
    if (typeof options === 'function') {
      cb = options;
    }

    var qs = {
      popreceipt: popReceipt,
      visibilitytimeout: visibilityTimeout
    };
    if (options.timeout) {
      qs.timeout = options.timeout;
    }

    this._makeRequest('PUT', queue+'/messages/'+messageId, qs, utils.xmlizeMessage.call(this, options || emptyObject, message), this._updateMessageCb.bind(this, cb));
    return this;
  },

  deleteMessage: function deleteMessage(queue, messageId, popReceipt, options, cb) {
    if (typeof options === 'function') {
      cb = options;
    }

    var qs = {
      popreceipt: popReceipt
    };
    if (options.timeout) {
      qs.timeout = options.timeout;
    }

    this._makeRequest('DELETE', queue+'/messages/'+messageId, qs, null, this._204Cb.bind(this, cb));
    return this;
  }
};

exports.Client = Client;